e2e/cypress/integration/notifications/desktop_notifications_spec.js

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+// ***************************************************************
+// - [#] indicates a test step (e.g. # Go to a page)
+// - [*] indicates an assertion (e.g. * Check the title)
+// - Use element ID when selecting an element. Create one if none.
+// ***************************************************************
+
+// Group: @notifications
+
+import * as MESSAGES from '../../fixtures/messages';
+import * as TIMEOUTS from '../../fixtures/timeouts';
+
+describe('Desktop notifications', () => {
+    let testTeam;
+    let testUser;
+
+    before(() => {
+        // Initialise a user.
+        cy.apiInitSetup({}).then(({team, user}) => {
+            testUser = user;
+            testTeam = team;
+        });
+    });
+
+    it('MM-T495 Desktop Notifications - Can set to DND and no notification fires on DM', () => {
+        cy.apiCreateUser({}).then(({user}) => {
+            cy.apiAddUserToTeam(testTeam.id, user.id);
+            cy.apiLogin(user);
+
+            cy.apiCreateDirectChannel([testUser.id, user.id]).then((res) => {
+                const channel = res.body;
+
+                // Mock window.Notification to check if desktop notifications are triggered.
+                cy.visit(`/${testTeam.name}/channels/town-square`, {
+                    onBeforeLoad(win) {
+                        cy.stub(win.Notification, 'permission', 'granted');
+                        cy.stub(win, 'Notification').as('Notification');
+                    },
+                });
```

Comment:
To ensure that the mocking works, we probably need to check the initial state where the user should receive a notification. With that, I'd suggest to move this into a function so we can reuse stubbing for with and without notification.
So for initial state, it can be,
```javascript
// # Ensure notifications are set up to fire a desktop notification if you receive a DM
cy.apiPatchUser(user.id, {notify_props: {...user.notify_props, desktop: 'all'}});

// Mock window.Notification to check if desktop notifications are triggered.
stubNotificationAs('withNotification');
cy.wait(TIMEOUTS.FIVE_SEC);

// # Have another user send you a DM
cy.postMessageAs({sender: testUser, message: MESSAGES.TINY, channelId: channel.id});

// * Desktop notification is received
cy.get('@withNotification').should('have.been.calledWithMatch', 'Direct Message', (args) => {
    expect(args.body).to.equal(`@${testUser.username}: ${MESSAGES.TINY}`);
    return true;
});
```

Then for `/dnd`
```javascript
stubNotificationAs('withoutNotification');
cy.wait(TIMEOUTS.FIVE_SEC);
...
cy.get('@withoutNotification').should('not.have.been.called');
```

Reply:
@saturninoabril makes sense, just wondering, do you see some way of injecting the notification stub outside of the `cy.visit` call like I've done it in the PR so far? Otherwise the `stubNotificationAs` function is going to have to do the `visit` call inside it I think.

Reply:
I'm good as is but maybe it can be done thru `cy.window`, like
```javascript
function stubNotificationAs(name) {
    cy.window().then((win) => {
        function Notification(title, opts) {
            this.title = title;
            this.opts = opts;
        }

        Notification.requestPermission = () => 'granted';
        Notification.close = () => true;

        win.Notification = Notification;

        cy.stub(win, 'Notification').as(name);
    });
}
```

Then, 
```javascript
cy.visit(`/${testTeam.name}/channels/town-square`);
stubNotificationAs('withoutNotification');
```

Reply:
I'll give it a try...

Reply:
@saturninoabril updated as per your suggestion.